parsers/JP.py

```diff
@@ -0,0 +1,132 @@
+# JP-KY  : Kyushu
+# JP-ON  : Okinawa
+
+def fetch_production(zone_key='JP-TK', target_datetime=None,
+                     logger=logging.getLogger(__name__)):
+    """
+    Calculates production from consumption and imports for a given area
+    All production is mapped to unknown
+    """
+    if target_datetime:
+        raise NotImplementedError(
+            'This parser is not yet able to parse past dates')
+    exch_map = {
+        'JP-HKD':['JP-TH'],
+        'JP-TH':['JP-TK'],
+        'JP-TK':['JP-TH', 'JP-CB'],
+        'JP-CB':['JP-TK', 'JP-HR', 'JP-KN'],
+        'JP-HR':['JP-CB', 'JP-KN'],
+        'JP-KN':['JP-CB', 'JP-HR', 'JP-SK', 'JP-CG'],
+        'JP-SK':['JP-KN', 'JP-CG'],
+        'JP-CG':['JP-KN', 'JP-SK', 'JP-KY']
+        }
+    df = fetch_consumption(zone_key, target_datetime)
+    df['imports'] = 0
+    for zone in exch_map[zone_key]:
+        df2 = occtonet.fetch_exchange(zone_key, zone, target_datetime)
+        df2 = pd.DataFrame(df2)
+        exchname = df2.loc[0, 'sortedZoneKeys']
+        df2 = df2[['datetime', 'netFlow']]
+        df2.columns = ['datetime', exchname]
+        df = pd.merge(df, df2, how='inner', on='datetime')
+        if exchname.split('->')[-1] == zone_key:
+            df['imports'] = df['imports']+df[exchname]
+        else:
+            df['imports'] = df['imports']-df[exchname]
+    df['prod'] = df['cons']-df['imports']
+    df = df[['datetime', 'prod']]
+    # add a row to production for each entry in the dictionary:
+    sources = {
+        'JP-HKD':'denkiyoho.hepco.co.jp',
+        'JP-TH':'setsuden.tohoku-epco.co.jp',
+        'JP-TK':'www.tepco.co.jp',
+        'JP-CB':'denki-yoho.chuden.jp',
+        'JP-HR':'www.rikuden.co.jp/denki-yoho',
+        'JP-KN':'www.kepco.co.jp',
+        'JP-SK':'www.energia.co.jp',
+        'JP-CG':'www.yonden.co.jp'
+        }
+    datalist = []
+    for i in range(df.shape[0]):
+        data = {
+            'zoneKey': zone_key,
+            'datetime': df.loc[i, 'datetime'],
+            'production': {
+                'biomass': None,
+                'coal': None,
+                'gas': None,
+                'hydro': None,
+                'nuclear': None,
+                'oil': None,
+                'solar': None,
+                'wind': None,
+                'geothermal': None,
+                'unknown': df.loc[i, 'prod']
+            },
+            'storage': {},
+            'source': ['occtonet.or.jp', sources[zone_key]]
+            }
+        datalist.append(data)
+    return datalist
+
+
+def fetch_consumption(zone_key='JP-TK', target_datetime=None,
+                      logger=logging.getLogger(__name__)):
+    """
+    Returns the consumption for an area as a pandas DataFrame
+    """
+    datestamp = arrow.get(target_datetime).to('Asia/Tokyo').strftime('%Y%m%d')
+    consumption_url = {
+        'JP-HKD': "".join(['http://denkiyoho.hepco.co.jp/area/data/juyo_01_',
+                           datestamp, '.csv']),
```

Comment:
for these you could do `'http://denkiyoho.hepco.co.jp/area/data/juyo_01_{}.csv'.format(datestamp)`